ui/lib/kv/addon/components/kv-list-filter.js

```diff
@@ -0,0 +1,154 @@
+ *
+ * @param {object} model - The adapter model object which contains an array of secret models.
+ * @param {string} mountPoint - Where in the router files we're located. For this component it will always be vault.cluster.secrets.backend.kv
+ * @param {string} filterValue - A concatenation between the list-directory's dynamic path "path-to-secret" and the queryParam "pageFilter". For example, if we're inside the beep/ directory searching for any secret that starts with "my-" this value will equal "beep/my-".
+ * @param {string} pageFilter - The queryParam value.
+ */
+
+export default class KvListFilterComponent extends Component {
+  @service router;
+  @tracked filterIsFocused = false;
+
+  kvRoute(route) {
+    return `${this.args.mountPoint}.${route}`;
+  }
+  /*
+  -partialMatch returns the secret that most closely matches the pageFilter queryParam.
+  -We're focused on pageFilter and not filterValue because if we're inside a directory we only care about the secrets listed there and not the directory. 
+  -If pageFilter is empty this returns the first secret model in the list.
+**/
+  get partialMatch() {
+    // you can't pass undefined to RegExp so if pageFilter is empty we replace it with an empty string.
+    const value = !this.args.pageFilter ? '' : this.args.pageFilter;
+    const reg = new RegExp('^' + escapeStringRegexp(value));
+    const match = this.args.model.secrets.filter((path) => reg.test(path.fullSecretPath))[0];
+    if (this.filterMatchesASecretPath || !match) return null;
+
+    return match.fullSecretPath;
+  }
+  /*
+  -filterMatchesASecretPath returns true if the filterValue matches a fullSecretPath
+  within the list of models.
+**/
+  get filterMatchesASecretPath() {
+    return !!this.args.model.secrets?.findBy('fullSecretPath', this.args.filterValue);
+  }
+  /*
+  -handleInput is triggered after the value of the input has changed. It is not triggered when input looses focus.
+**/
+  @action
+  handleInput(event) {
+    const input = event.target.value;
+    const isDirectory = keyIsFolder(input);
+    const parentDirectory = parentKeyForKey(input);
+    const secretWithinDirectory = keyWithoutParentKey(input);
+    // TODO kv engine cleanup: ideally when it's not a directory we could filter through the current models and remove pageFilter refresh on the list route.
+    if (isDirectory) {
+      this.router.transitionTo(this.kvRoute('list-directory'), input);
+    } else if (parentDirectory) {
+      this.router.transitionTo(this.kvRoute('list-directory'), parentDirectory, {
+        queryParams: { pageFilter: secretWithinDirectory },
+      });
+    } else {
+      this.router.transitionTo(this.kvRoute('list'), {
+        queryParams: { pageFilter: input },
+      });
+    }
+  }
+  /*
+  -handleKeyDown handles: tab, enter, backspace and escape. Ignores everything else.
+**/
+  @action
+  handleKeyDown(event) {
```

Comment:
This method feels pretty overloaded with logic. I wonder if we can split out the specific key handling to other methods (`handleTab`, `handleEnter`) etc and this method would only be in charge of checking if the key event is one of the special keys, and preventing the default behavior if it is

Reply:
Good idea. I'll amend.